src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java

```diff
@@ -0,0 +1,151 @@
+import java.io.File;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.Proxy;
+import java.net.URL;
+
+import static org.apache.commons.lang3.StringUtils.defaultIfBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.jenkinsci.plugins.github.config.GitHubServerConfig.GITHUB_URL;
+import static org.jenkinsci.plugins.github.config.GitHubServerConfig.tokenFor;
+
+/**
+ * Converts server config to authorized GH instance on {@link #applyNullSafe(GitHubServerConfig)}.
+ * If login process is not successful it returns null
+ *
+ * Uses okHttp (https://github.com/square/okhttp) as connector to have ability to use cache and proxy
+ * The capacity of cache can be changed with help of {@link GitHubLoginFunction#CACHE_SIZE_MB_FOR_GITHUB} as
+ * system property.
+ *
+ * Don't use this class in any place directly
+ * as of it have public static factory {@link GitHubServerConfig#loginToGithub()}
+ *
+ * @author lanwen (Merkushev Kirill)
+ * @see GitHubServerConfig#loginToGithub()
+ */
+@Restricted(NoExternalUse.class)
+public class GitHubLoginFunction extends NullSafeFunction<GitHubServerConfig, GitHub> {
+
+    /**
+     * Capacity of cache for GitHub client in MB.
+     *
+     * Can be overridden as system property -Dorg.jenkinsci.plugins.github.GitHubPlugin.cache.size.mb=1 on jenkins jvm
+     * Set to <= 0 to turn off this feature
+     *
+     * Defaults to 20 MB
+     *
+     * @since TODO
+     */
+    private static final int CACHE_SIZE_MB_FOR_GITHUB = Integer.getInteger(
+            GitHubPlugin.class.getName() + ".cache.size.mb", 20
+    );
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GitHubLoginFunction.class);
+
+    /**
+     * Called by {@link #apply(Object)}
+     * Logins to GH and returns client instance
+     *
+     * @param github config where token saved
+     *
+     * @return authorized client or null on login error
+     */
+    @Override
+    @CheckForNull
+    protected GitHub applyNullSafe(@Nonnull GitHubServerConfig github) {
+        String accessToken = tokenFor(github.getCredentialsId());
+
+        GitHubBuilder builder = new GitHubBuilder()
+                .withOAuthToken(accessToken)
+                .withConnector(connector(defaultIfBlank(github.getApiUrl(), GITHUB_URL)))
+                .withRateLimitHandler(RateLimitHandler.FAIL);
+        try {
+            if (isNotBlank(github.getApiUrl())) {
+                builder.withEndpoint(github.getApiUrl());
+            }
+
+            return builder.build();
+        } catch (IOException e) {
+            LOGGER.warn("Failed to login with creds {}", github.getCredentialsId(), e);
+            return null;
+        }
+    }
+
+    /**
+     * Uses proxy if configured on pluginManager/advanced page
+     *
+     * @param apiUrl GitHub's url to build proxy to
+     *
+     * @return proxy to use it in connector
+     */
+    private Proxy getProxy(String apiUrl) {
```

Comment:
It for extra checks for null when use something. If method private and used in one place - its overhead to annotate methods with nonnull